### PR TITLE
Dts tests

### PR DESCRIPTION
**Pull request by macpijan**
Somewhat based on: https://github.com/Dasharo/open-source-firmware-validation/pull/267

Requires: https://github.com/Dasharo/meta-dts/pull/136

Test output:

```
robot -L TRACE -v config:qemu -v rte_ip:127.0.0.1 -v snipeit:no dts/dts-tests.robot
==============================================================================
Dts-Tests                                                                     
==============================================================================
PC Engines DES initial deployment (legacy -> UEFI) - no credential... | PASS |
------------------------------------------------------------------------------
PC Engines DES initial deployment (legacy -> UEFI) :: Verify DES (... | PASS |
------------------------------------------------------------------------------
PC Engines DES initial deployment (legacy -> SeaBIOS) :: Verify DE... | FAIL |
DES SeaBIOS must be supported in DTS first
------------------------------------------------------------------------------
NCM transition (UEFI -> heads) - elgible UEFI version + heads cred... | PASS |
------------------------------------------------------------------------------
NCM transition (UEFI -> heads) - elgible UEFI version + no credent... | PASS |
------------------------------------------------------------------------------
Dts-Tests                                                             | FAIL |
5 tests, 4 passed, 1 failed
==============================================================================
Output:  /home/macpijan/projects/github/dasharo/open-source-firmware-validation/output.xml
Log:     /home/macpijan/projects/github/dasharo/open-source-firmware-validation/log.html
Report:  /home/macpijan/projects/github/dasharo/open-source-firmware-validation/report.html
```

### CODE REVIEW

**Comment by DaniilKl** [comment]
Some rebasing on latest develop.

File: dts/dts-tests.robot [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+*** Test Cases ***
+PC Engines DPP initial deployment (legacy -> UEFI) - no credentials
+    [Documentation]    Verify DPP (coreboot + UEFI) initial deployment logic on
+    ...    PC Engines. We start from legacy firmware and insert no DPP keys. No
+    ...    firmware shall be offered, but we expect link to shop.
+
+    Power On And Enter DTS Shell
+
+    Execute Command In Terminal
+    ...    export BOARD_VENDOR="PC Engines" SYSTEM_MODEL="APU2" BOARD_MODEL="APU2" SYSTEM_VENDOR="PC Engines"
```
Comment (DaniilKl) [inline dts/dts-tests.robot:35]: I am not sure how it works. Ok, you will set globally these variables before executing DTS scripts, but the variables are being overwritten in `dts-environment.sh`, which is being sourced in every script.

File: dts/dts-tests.robot [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+    ${out}=    Read From Terminal Until    Does it match your actual specification? (Y|n)
+    Write Into Terminal    Y
+    Log    ${out}
+
+    ${out}=    Read From Terminal Until    Do you want to install Dasharo firmware on your hardware? (Y|n)
+    Write Into Terminal    Y
+    Log    ${out}
+
+    ${out}=    Read From Terminal Until    Rebooting
```
Comment (DaniilKl) [inline dts/dts-tests.robot:82]: Will it get here? In DTS scripts, you are communicating with hardware not only via `dmidecode` in `dts-environment.sh` but in many other ways, including `flashrom` read/write commands, and all of them will fail on QEMU. So, will the test get here if you will launch it on QEMU?

**Review by DaniilKl — COMMENTED** [review]
Additionally, I have problems running this tests:

```bash
λ robot -b command_log.txt -v snipeit:no -L TRACE -v config:qemu -t "*" dts/dts-tests.robot
[ ERROR ] Error in file '/home/danillklimuk/Projects/DTS/open-source-firmware-validation/keywords.robot' on line 4: Resource file 'pikvm-rest-api/pikvm_comm.robot' does not exist.
[ ERROR ] Error in file '/home/danillklimuk/Projects/DTS/open-source-firmware-validation/dts/dts-tests.robot' on line 11: Resource file '../rtectrl-rest-api/rtectrl.robot' does not exist.
[ ERROR ] Error in file '/home/danillklimuk/Projects/DTS/open-source-firmware-validation/dts/dts-tests.robot' on line 12: Resource file '../sonoff-rest-api/sonoff-api.robot' does not exist.
==============================================================================
Dts-Tests                                                                     
==============================================================================
PC Engines DPP initial deployment (legacy -> UEFI) - no credential... | FAIL |
Parent suite setup failed:
Evaluating IF condition failed: Variable '${RTE_IP}' not found. Did you mean:
    ${RTE_LIST}
------------------------------------------------------------------------------
PC Engines DPP initial deployment (legacy -> UEFI) :: Verify DPP (... | FAIL |
Parent suite setup failed:
Evaluating IF condition failed: Variable '${RTE_IP}' not found. Did you mean:
    ${RTE_LIST}
------------------------------------------------------------------------------
(...)
```

I guess it is caused by latest changes related to submodules.

**Comment by DaniilKl** [comment]
> I guess it is caused by latest changes related to submodules. 

Deleted this submodules here 8c726ad9cd0990a526461b13af7922dccd526fe1, and added example robot command for QEMU here 7b1f7591953c7819a4d8664158e3363e00040d5d, in case smbd will not know how to run it :sweat_smile:.

File: dts/dts-tests.robot (hunk 1 above)
Comment (DaniilKl) [reply dts/dts-tests.robot:33]: I see, that in `dts-environment` some variables assignment checks if the variables already has a value:

```bash
BIOS_VENDOR="${BIOS_VENDOR:-$(dmidecode -s bios-vendor)}"
BIOS_VERSION="${BIOS_VERSION:-$(dmidecode -s bios-version)}"
``` 

So, this question is closed.

File: dts/dts-tests.robot (hunk 2 above)
Comment (DaniilKl) [reply dts/dts-tests.robot:80]: And here is a confirmation. I have been trying to go through your second test (which is `PC Engines DPP initial deployment (legacy -> SeaBIOS)`), and it did not pass:

```bash
==============================================================================
Dts-Tests                                                                     
==============================================================================
PC Engines DPP initial deployment (legacy -> UEFI) :: Verify DPP (... | FAIL |
No match found for 'Rebooting' in 6 minutes 30 seconds. Output:

Backing up BIOS firmware and store it locally...
Remember that firmware is also backed up in HCL report.
Failed to read BIOS firmware backup : (1)
Press any key to continue.
------------------------------------------------------------------------------
```

As you can see it has failed on one of the `flashrom` commands ([here](https://github.com/Dasharo/dts-scripts/blob/5ebb3d910764a4589c6473a577c822967b3b0d19/scripts/dasharo-deploy#L207)), because `flashrom` does not work with QEMU this could be considered as false negative.

Comment (m-iwanicki) [reply dts/dts-tests.robot:80]: Shouldn't `$FLASHROM` be set to https://github.com/Dasharo/dts-scripts/blob/main/tests/flashrom-mock.sh? That's how these tests work, by either setting variables before (so `dmidecode` isn't used) or by using mock/fake version of tool

Comment (m-iwanicki) [reply dts/dts-tests.robot:80]: `FLASHROM="${FLASHROM:-flashrom}"` this will set `FLASHROM` to `flashrom` only if variable isn't already set. Maybe you missed some configuration/preparation step?

Comment (macpijan) [reply dts/dts-tests.robot:80]: Exactly. See: https://github.com/Dasharo/dts-scripts/tree/main/tests#running-manually and https://github.com/Dasharo/dts-scripts/blob/main/tests/dts-boot

Comment (DaniilKl) [reply dts/dts-tests.robot:80]: It should, but do we have all needed mock/fake versions now? What about handling calls to tools, which read hardware inf., e.g. [during configuration](https://github.com/Dasharo/dts-scripts/blob/5ebb3d910764a4589c6473a577c822967b3b0d19/include/dts-functions.sh#L371)

Comment (macpijan) [reply dts/dts-tests.robot:80]: We have what can be seen. So flashrom end ectool, and that was enough to run the scenarios shown in the first comment at the time of writing this.

We never know if we will not execute these (or other) tests any longer.

The linked code is wrong, it should use variable not command directly. Clearly it will cause troubles in qemu.

Comment (DaniilKl) [reply dts/dts-tests.robot:80]: :writing_hand:

Comment (m-iwanicki) [reply dts/dts-tests.robot:80]: We could prepend `PATH`  so it contains our `fake` tools, just rename to correct names e.g. `flashrom-mock.sh` to `flashrom`. Linux should use tool from first available directory in `PATH`

**Comment by DaniilKl** [comment]
First test works:

```bash
==============================================================================
Dts-Ptp                                                                       
==============================================================================
PTP001.001 PC Engines DPP initial deployment (legacy -> Dasharo) -... | PASS |
------------------------------------------------------------------------------
Dts-Ptp                                                               | PASS |
1 test, 1 passed, 0 failed
==============================================================================
```

**Comment by DaniilKl** [comment]
The PTP tests work. How to use:

```bash
λ robot -b command_log.txt -v snipeit:no -L TRACE -v config:qemu -v rte_ip:127.0.0.1 -v netboot_utilities_support:True -v dts_ipxe_link:http://192.168.0.104:8080/ipxe -v dpp_logs_key:'x' -v dpp_download_key:'x' -v dpp_password:'x' dts/dts-ptp.robot
```

* `dts_ipxe_link`: useful if you are testing DTS which is not released yet. Just put here a link to your script which will load your DTS. By default DTS is being booted from `dl.3mdeb.com`;
* `dpp_logs_key`, `dpp_download_key`, `dpp_password`: for DPP credentials, if tests need them.

Issues:

```bash
robocop..................................................................Failed
- hook id: robocop
- exit code: 2

/home/danillklimuk/Projects/DTS/open-source-firmware-validation/dts/dts-ptp.robot:413:1 [W] 0504 Test case 'PTP002.009 NCM V540TNC_TND_TNE installation (legacy -> Coreboot + UEFI) - community version' is too long (58/50) (too-long-test-case)
/home/danillklimuk/Projects/DTS/open-source-firmware-validation/dts/dts-ptp.robot:473:1 [W] 0504 Test case 'PTP002.010 NCM V560TNC_TND_TNE installation (legacy -> Coreboot + UEFI) - community version' is too long (59/50) (too-long-test-case)
```

I suppose I left too many comments in these test cases... Actually this tests suite needs some polishing.

**Comment by DaniilKl** [comment]
```bash
robocop..................................................................Failed
- hook id: robocop
- exit code: 3

/code/lib/dts-lib.robot:36:1 [W] 0501 Keyword 'Boot Dasharo Tools Suite' is too long (51/50) (too-long-keyword)
/code/dts/dts-ptp.robot:413:1 [W] 0504 Test case 'PTP002.009 NCM V540TNC_TND_TNE initial deployment (legacy -> Coreboot + UEFI) - community version' is too long (58/50) (too-long-test-case)
/code/dts/dts-ptp.robot:473:1 [W] 0504 Test case 'PTP002.010 NCM V560TNC_TND_TNE initial deployment (legacy -> Coreboot + UEFI) - community version' is too long (59/50) (too-long-test-case)
```

Is there any reasons we should adhere to these checks? I prefer to use comments as much as I can, this is why I get these errors.

**Comment by DaniilKl** [comment]
The example of these tests being used can be found here https://github.com/Dasharo/dts-scripts/pull/26.

File: docs/dts-tests.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# DTS tests
+
+This document describes current DTS tests coverage. Tests planned:
+
+* PTP tests;
+* Unit tests.
+
+## PTP tests
+
+PTP - Point To Point tests, tests that verify how DTS goes from start to end
```
Comment (macpijan) [inline docs/dts-tests.md:10]: Doesn't E2E fit better here?

Comment (DaniilKl) [reply docs/dts-tests.md:10]: Perhaps :thinking:

Comment (DaniilKl) [reply docs/dts-tests.md]: replaced here 456660150dedb8709e8ba5886630c7255d065741.

File: docs/dts-tests.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+every user workflow (e.g. installation, update, etc.) for every platform.
+
+Location in OSFV: `dts/dts-e2e.robot`.
+
+These tests include modifications for DTS platform emulation so the tests
+could be launched on Qemu. Then every test case choose the workflow by
+choosing DTS menu option, and provides credentials (if necessary). So, the
+start conditions are: platform configuration and workflow selection.
+
+Then every test cases go through the chosen workflow and checks for expected
```
[the remark this answers is not in the document]
Comment (DaniilKl) [reply docs/dts-tests.md]: Fixed [here](https://github.com/Dasharo/open-source-firmware-validation/compare/456660150dedb8709e8ba5886630c7255d065741..ff166a65e8675c5dbcd984180835f72a6b9b8e3d)

File: docs/qemu.md [hunk 5]
```diff
@@ -47,3 +47,16 @@ to see which test have been proven to work on QEMU so far.
...
 You may also refer to the `./scripts/ci/qemu-self-test.sh`, where we aim to
 keep testing common keywords, to ensure of their correct operation.
+
+## Running tests
+
+Here is an example command for running tests on QEMU (make sure that you are in
+Python virtual environment):
+
+```bash
+robot -b command_log.txt -v snipeit:no -L TRACE -v config:qemu -v rte_ip:127.0.0.1 -t "*" dts/dts-tests.robot
```
[the remark this answers is not in the document]
Comment (DaniilKl) [reply docs/qemu.md]: Fixed [here](https://github.com/Dasharo/open-source-firmware-validation/compare/456660150dedb8709e8ba5886630c7255d065741..ff166a65e8675c5dbcd984180835f72a6b9b8e3d)

**Review by PLangowski — APPROVED** [review]

**Comment by artur-rs** [comment]
@DaniilKl on the weekly meeting, we've agreed to merge these changes when:
* some logs from tests will be provided
* merge conflict will be resolved

**Comment by DaniilKl** [comment]
@artur-rs, as I said on one of our meetings, this PR needs small changes after review in https://github.com/Dasharo/dts-scripts/pull/26. Changing it to Draft.

**Review by m-iwanicki — CHANGES_REQUESTED** [review]

File: lib/dts-lib.robot [hunk 6]
```diff
@@ -90,34 +97,6 @@ Enter Shell In DTS
     Log    ${out}
     Remove Extra Default Route
 
-Run EC Transition
```
[the remark this answers is not in the document]
Comment (DaniilKl) [reply lib/dts-lib.robot:93]: dfaa07f0a46d1271aaf71159219c05d500fb4b3b

[the remark this answers is not in the document]
Comment (m-iwanicki) [reply lib/dts-lib.robot:93]: @macpijan Is removing this test in this case ok?
What about https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/326-dasharo-tools-suite/#dts008001-flash-device-ec-firmware-by-using-dts-built-in-script-works-correctly?
And question about those docs as they are a little out of date:

1. Which branch/tag/release does this documentation relate to?
2. Does `EC Transition == EC installation` (moving from proprietary to Dasharo EC)?

**Review by PLangowski — APPROVED** [review]

File: lib/dts-lib.robot (hunk 6 above)
[the remark this answers is not in the document]
Comment (DaniilKl) [reply lib/dts-lib.robot]: Reverted these changes [here](https://github.com/Dasharo/open-source-firmware-validation/compare/81717571be9efb2a00ad57f4f6c5469e6a3f427b..86ae00c85976d14449cee8e0b98d15fbb88cee2a).

**Review by m-iwanicki — APPROVED** [review]

**Comment by DaniilKl** [comment]
Rebased on the latest develop.

**Comment by PLangowski** [comment]
@DaniilKl Could you also add tags to the test cases that match the DPP subscription?
For example:

```
E2E003.011 MSI PRO Z690-A DDR4 transition (Coreboot + UEFI -> heads) - without credentials
[tags] msi_heads
```

It would make it much easier to run all tests for a given set of credentials

File: docs/dts-tests.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+Then every test case goes through the chosen workflow and checks for expected
+behavior. If everything goes as expected - test case passes, otherwise - it
+fails.
+
+Control variables:
+
+* `dts_ipxe_link`: useful if you are testing DTS which is not released yet. Just
+  put here a link to your script which will load your DTS. By default DTS is
+  being booted from `dl.3mdeb.com`;
+* `dpp_logs_key`, `dpp_download_key`, `dpp_password`: for DPP credentials, if
```
Comment (macpijan) [inline docs/dts-tests.md:29]: Can we apply this suggestion here: https://github.com/Dasharo/meta-dts/pull/193/files#r1853692417 ?
Let me know your thoughts on this.

File: docs/dts-tests.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  tests need them.
+
+Launching example:
+
+```bash
+robot -b command_log.txt -v snipeit:no -L TRACE -v config:qemu -v rte_ip:127.0.0.1 -v netboot_utilities_support:True -v dts_ipxe_link:http://192.168.0.102:8080/ipxe -v dpp_logs_key:'LOGS_KEY' -v dpp_download_key:'DOWNLOAD_KEY' -v dpp_password:'PASSWORD' -t "E2E006.002*" dts/dts-e2e.robot
```
Comment (macpijan) [inline docs/dts-tests.md:35]: `-v netboot_utilities_support:True` - why is this needed?

Comment (macpijan) [reply docs/dts-tests.md:35]: This suite https://github.com/Dasharo/open-source-firmware-validation/blob/develop/dasharo-compatibility/network-boot-utilities.robot is specific to Protectli platforms

**Comment by macpijan** [comment]
> Could you also add tags to the test cases that match the DPP subscription?

Tags can be added as well - it does not hurt to add them, to have the ability to quickly test only a subset of the regression suite.

But for the full regression, I think we want to run the whole file in one run, perhaps with such modification: https://github.com/Dasharo/open-source-firmware-validation/pull/333/files#r1853695808

File: docs/dts-tests.md (hunk 8 above)
Comment (m-iwanicki) [reply docs/dts-tests.md:35]: https://github.com/Dasharo/open-source-firmware-validation/blob/b40b2f273b8866917398aef8ccbf1b990717724a/lib/dts-lib.robot#L51
From what I understand so we can boot custom DTS via iPXE shell

Comment (macpijan) [reply docs/dts-tests.md:35]:
```
    ELSE IF    '${dts_booting_method}'=='iPXE'
        IF    ${NETBOOT_UTILITIES_SUPPORT} == ${TRUE}
            # DTS_IPXE_LINK can be defined before running tests, e.g. via CMD or
            # some file:
            Boot Dasharo Tools Suite Via IPXE Shell    ${DTS_IPXE_LINK}
        ELSE
            ${boot_menu}=    Enter Boot Menu Tianocore And Return Construction
            Enter Submenu From Snapshot    ${boot_menu}    ${IPXE_BOOT_ENTRY}
            ${ipxe_menu}=    Get IPXE Boot Menu Construction
            Enter Submenu From Snapshot    ${ipxe_menu}    Dasharo Tools Suite
            Set DUT Response Timeout    5m
            Read From Terminal Until    .cpio.gz...
            Read From Terminal Until    ok
        END
    ELSE
        FAIL    Unknown or improper connection method: ${dts_booting_method}
    END
```

The `NETBOOT_UTILITIES_SUPPORT` variable is set for Protectli devices only, as they have a bit different iPXE menu. According to this code, the `ELSE` branch is supposed to be used for boards other than `Protectli` (with regular Dasharo iPXE menu).

If the first `IF` branch works for both `Protectli` and `Dasharo` iPXE menu (on QEMU we have Dasharo menu), then maybe we do not need this if at all, and we can always use the code from `IF` branch.

Comment (macpijan) [reply docs/dts-tests.md:35]: TLDR: if the current code works for you, you can remove this `IF statement` to always select first branch, and not set the NETBOOT_UTILITIES_SUPPORT variable.

File: docs/dts-tests.md (hunk 7 above)
Comment (DaniilKl) [reply docs/dts-tests.md:29]: This will be a problem for a short period of time, because soon, when we migrate our infrastructure to MinIO, we will be able to use one set of credentials for all subscriptions. So, I would rather stay with the way it is now.

File: docs/dts-tests.md (hunk 8 above)
Comment (DaniilKl) [reply docs/dts-tests.md:35]: Both work on qemu, but the bootflow after `ELSE` boots DTS from boot.dasharo.com using menu options, when the bootflow after `IF` uses iPXE shell, that  enables booting DTS from custom link.

I gues I should rename `NETBOOT_UTILITIES_SUPPORT` variable then.

Comment (DaniilKl) [reply docs/dts-tests.md]: Changed here cc392489aae15f9a1b8764a205ae48cce4c22839.

**Comment by DaniilKl** [comment]
Added tags here be03f347bc525cf96d2df3062466fac08489a04f.

File: lib/dts-lib.robot [hunk 9]
```diff
@@ -2,18 +2,54 @@
 Resource    terminal.robot
 
 
+*** Variables ***
+# Default DTS link for iPXE boot, can be overwritten by CMD:
+${DTS_IPXE_LINK}=               http://boot.dasharo.com/dts/dts.ipxe
+${BOOT_DTS_FROM_IPXE_SHELL}=    ${FALSE}
```
Comment (m-iwanicki) [inline lib/dts-lib.robot:8]: Shouldn't this be defined in `platform-configs/include/default.robot`?

File: lib/dts-lib.robot [hunk 10]
```diff
@@ -2,18 +2,54 @@
...
+    # 3) Set up net card:
     Write Into Terminal    dhcp net0
     ${out}=    Read From Terminal Until Prompt
     Should Contain    ${out}    ok
     Set DUT Response Timeout    60s
-    Write Bare Into Terminal    chain http://boot.dasharo.com/dts/dts.ipxe\n    0.1
-    Read From Terminal Until    http://boot.dasharo.com/dts/dts.ipxe...
+
+    # 4) Try to boot via the link:
+    Write Bare Into Terminal    chain ${dts_chain_link}\n
```
Comment (m-iwanicki) [inline lib/dts-lib.robot:32]: Why did you remove interval argument (`0.1`)? Won't this affect another tests?

File: lib/dts-lib.robot (hunk 9 above)
Comment (DaniilKl) [reply lib/dts-lib.robot]: Agreed, changed [here](https://github.com/Dasharo/open-source-firmware-validation/compare/be03f347bc525cf96d2df3062466fac08489a04f..18a702db97333836ce2084ea95643c7a74e78e78).

File: lib/dts-lib.robot (hunk 10 above)
Comment (DaniilKl) [reply lib/dts-lib.robot]: Good catch, missed this one, fixed [here](https://github.com/Dasharo/open-source-firmware-validation/compare/be03f347bc525cf96d2df3062466fac08489a04f..18a702db97333836ce2084ea95643c7a74e78e78)

File: docs/dts-tests.md (hunk 7 above)
Comment (macpijan) [reply docs/dts-tests.md:29]: ok, let's see how short this period will be then

File: docs/dts-tests.md (hunk 8 above)
Comment (macpijan) [reply docs/dts-tests.md]: Ok I get it why we need to boot from shell now. Let's stay with this approach for now.

**Comment by macpijan** [comment]
@DaniilKl Cannot approve as you took over my MR - which is fine.

Approving here, we can move forward.
